### PR TITLE
Add error handling

### LINKED ISSUE #3: Handle errors gracefully

### PR DESCRIPTION
**Pull request by FlakM**
The style of error handling was inspired by https://michael-f-bryan.github.io/rust-ffi-guide/errors/index.html it holds errors in local thread storage and exposes functions `last_error_message` and `last_error_length` to access the messages over FFI. All functions return POSIX style return codes:

- `null` for references
- `-1` for longs and lengths

There are couple of things to consider:

- [ ] making it completely panic free by catching all errors using `std::panic::catch_unwind` checkout https://doc.rust-lang.org/nomicon/ffi.html#panic-can-be-stopped-at-an-abi-boundary but this seems like an overkill for non safety critical app
- [ ] if there is any possibility that some code will ever check the error messages from other thread than the one that resulted in the problem itself?

Closes #3

### CODE REVIEW

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
// Return a pointer to encapsulated request
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Return the size in bytes of the encapsulated request.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Return a pointer to the decapsulated response.
///
/// # Safety
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Return size in bytes of the decapsulated response.
///
/// # Safety
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Encapsulates the provided `encoded_msg` using `encoded_config` and return
/// a context used for decapsulating the corresponding response.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// This function will return a NULL pointer if:
///     - creating the request context fails due to input errors.
///     - encapsulation fails.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/lib.rs
Comment (chris-wood) [inline apprelay/src/lib.rs]:
```suggestion
/// Decapsulates the provided `encapsulated_response` using `context`.
///
/// This function will return a NULL pointer if decapsulation fails.
///
/// # Safety
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/error_ffi.rs
Comment (chris-wood) [inline apprelay/src/error_ffi.rs]:
```suggestion
/// Otherwise the function returns the number of bytes written to buffer.
```

File: apprelay/src/error_ffi.rs
Comment (chris-wood) [inline apprelay/src/error_ffi.rs]:
```suggestion
/// Return the number of bytes in the last error message.
/// Does not include any trailing null terminators.
#[no_mangle]
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// Decapsulates the provided response `encapsulated_response` using
/// requests config obtain by dereferencing `context_ptr`
///
/// Returns an array containing the decapsulated response.
///
/// If this function fails due JNI problems or decapsulation it returns a NULL pointer.
///
/// # Safety
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// This dereferences a raw pointer to `RequestContext` passed by the caller.
/// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// Accesses the encapsulation result for given context.
///
/// Returns an array containing an encapsulation request,
/// or -1 upon failure.
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// Encapsulates a request using the provided configuration.
///
/// Returns a pointer to encapsulation context, and returns -1 upon failure.
```

File: apprelay/src/android.rs
Comment (chris-wood) [inline apprelay/src/android.rs]:
```suggestion
/// Return most recent error as a Java `String`.
///
/// If the are no recent errors than this returns a NULL pointer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Otherwise the function returns the number of bytes written to buffer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Decapsulates the provided `encapsulated_response` using `context`.
//
// This function will return a NULL pointer if decapsulation fails.
//
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// This function will return a NULL pointer if:
//     - creating the request context fails due to input errors.
//     - encapsulation fails.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Return the size in bytes of the decapsulated response.
//
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Return a pointer to the decapsulated response.
//
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
//
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Return the size in bytes of the encapsulated request.
//
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
```

File: apprelay/apprelay.h
Comment (chris-wood) [inline apprelay/apprelay.h]:
```suggestion
// Return the pointer to first byte of encapsulated request.
//
// # Safety
// This dereferences a raw pointer to `RequestContext` passed by the caller.
// Be sure that the context has not been yet freed and that you are using valid pointer.
```

**Review by chris-wood — APPROVED** [review]
LGTM pending some documentation improvements. I don't think we need to make this sample thread safe. We should probably just note that it's not thread safe. As for the catch-all panic change, is that something we could do after landing this PR?

**Comment by FlakM** [comment]
I guess catch unwinding might be necessary in context of encapsulation/decapsulation if the code might possibly panic there, do you think it is possible?

**Comment by chris-wood** [comment]
> I guess catch unwinding might be necessary in context of encapsulation/decapsulation if the code might possibly panic there, do you think it is possible?

It doesn't look like there are panics on the code paths we use in the `ohttp` crate, but it is probably a good thing to catch them just in case. Can we do this in a followup change?

**Pull request merged**
